src/Open3D/GUI/FileDialog.cpp

```diff
@@ -0,0 +1,444 @@
+// ----------------------------------------------------------------------------
+// -                        Open3D: www.open3d.org                            -
+// ----------------------------------------------------------------------------
+// The MIT License (MIT)
+//
+// Copyright (c) 2018 www.open3d.org
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+// IN THE SOFTWARE.
+// ----------------------------------------------------------------------------
+
+#if defined(__APPLE__)
+// see FileDialogNative.cpp
+#else
+
+#include "FileDialog.h"
+
+#include "Button.h"
+#include "Combobox.h"
+#include "Label.h"
+#include "Layout.h"
+#include "ListView.h"
+#include "TextEdit.h"
+#include "Theme.h"
+#include "Util.h"
+
+#include "Open3D/Utility/Console.h"
+#include "Open3D/Utility/FileSystem.h"
+#include "Open3D/Utility/Helper.h"
+
+#include <string>
+#include <unordered_map>
+#include <unordered_set>
+
+// macOS sorts directories in with the files
+// Windows and Linux (GTK) sort directories first.
+#ifdef __APPLE__
+#define INLINE_DIRS 1
+#else
+#define INLINE_DIRS 0
+#endif  // __APPLE__
+
+namespace open3d {
+namespace gui {
+
+namespace {
+// The current path of the dialog should persist across runs of the dialog.
+// This is defined here rather than in the class definition because we can't
+// need to be exporting internal details of the class.
+static std::string gFileDialogDir;
+}  // namespace
+
+class DirEntry {
+public:
+    enum Type { DIR, FILE };
```

Comment:
That is a good point. I've changed FileDialog::Type to FileDialog::Mode, since that seems a little more appropriate anyway. Please unresolve the conversation if you don't think that's a good idea.